src/s3ql/http.py

```diff
@@ -0,0 +1,1558 @@
+
+        If *substr* cannot be found in the next *maxsize* bytes,
+        raises `_ChunkTooLong`.
+        '''
+
+        if not isinstance(substr, (bytes, bytearray, memoryview)):
+            raise TypeError('*substr* must be bytes-like')
+
+        log.debug('reading until %s', substr)
+
+        rbuf = self._rbuf
+        sub_len = len(substr)
+
+        # Make sure that substr cannot be split over more than one part
+        assert len(rbuf.d) > sub_len
+
+        parts = []
+        while True:
+            # substr may be split between last part and current buffer
+            # This isn't very performant, but it should be pretty rare
+            if parts and sub_len > 1:
+                buf = b''.join(
+                    (parts[-1][-sub_len:], rbuf.d[rbuf.b : min(rbuf.e, rbuf.b + sub_len - 1)])
+                )
+                idx = buf.find(substr)
+                if idx >= 0:
+                    idx -= sub_len
+                    break
+
+            # log.debug('rbuf is: %s', rbuf.d[rbuf.b:min(rbuf.e, rbuf.b+512)])
+            stop = min(rbuf.e, rbuf.b + maxsize)
+            idx = rbuf.d.find(substr, rbuf.b, stop)
+
+            if idx >= 0:  # found
+                break
+            if stop != rbuf.e:
+                raise _ChunkTooLong()
+
+            # If buffer is full, store away the part that we need for sure
+            if rbuf.e == len(rbuf.d):
+                log.debug('buffer is full, storing part')
+                buf = rbuf.exhaust()
+                parts.append(buf)
+                maxsize -= len(buf)
+
+            # Refill buffer
+            while True:
+                res = self._try_fill_buffer()
+                if res is None:
+                    log.debug('need more data, yielding')
+                    await PollNeeded(self._sock.fileno(), POLLIN)
+                elif res == 0:
+                    raise ConnectionClosed('server closed connection')
+                else:
+                    break
+
+        log.debug('found substr at %d', idx)
+        idx += len(substr)
+        buf = rbuf.d[rbuf.b : idx]
+        rbuf.b = idx
+
+        if parts:
+            parts.append(buf)
+            buf = b''.join(parts)
+
+        try:
+            return buf.decode('latin1')
+        except UnicodeDecodeError:
+            raise InvalidResponse('server response cannot be decoded to latin1')
+
+    def _try_fill_buffer(self):
+        '''Try to fill up read buffer
+
+        Returns the number of bytes read into buffer, or `None` if no data was
+        available on the socket. Return zero if the TCP connection has been
+        properly closed but the socket object still exists. On other problems
+        (e.g. if the socket object has been destroyed or the connection
+        interrupted), raise `ConnectionClosed`.
+        '''
+
+        log.debug('start')
+        rbuf = self._rbuf
+
+        # If buffer is empty, reset so that we start filling from beginning
```

Comment:
Is there a reason the buffer handling like this isn't moved into the buffer class?